client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -135,30 +136,37 @@ protected Compressor initialValue() {
 
   private final ReviveManager reviveManager;
 
+  private final boolean dataPushFailureTrackingEnabled;
+
   protected static class ReduceFileGroups {
     public Map<Integer, Set<PartitionLocation>> partitionGroups;
+    public Map<String, Set<PushFailedBatch>> pushFailedBatches;
```

Comment:
QQ: This is essentially relevant only when hard_split's happen, right ?
For all other cases, either we retry and it succeeds or it fails ?
I want to make sure I am not missing some other cases.

Reply:
Yes, maybe we retry push data, but maybe primary peer success but replica peer failed, we want to exclude this batch to avoid reading data repeatly

Reply:
The concern I have with this is, the data would be under replicated, right ?
If yes, wouldn't it not be simply better to retry with an updated peer ?

If that is an option, the problem you [described above](https://github.com/apache/celeborn/pull/2373#discussion_r1562166228) is designed away as well ?

Reply:
I'm not quite sure if I fully understand your meaning, do you mean: 
- `the data would be under replicated`, this means data will be lost ? I think data will not be lost, because failed batch will be retried to another PartitionLocation requested by Revive with epoch + 1 
- `If yes, wouldn't it not be simply better to retry to the same (or perhaps updated) peer ?` Revive is designed to avoid retry push data to the same PartitionLocation, maybe sometimes retry the same node will not success or timeout. And we want to ensure that one Succeed batch should be available both Primary and Replicate Peer

If I misunderstood, could you kindly explain further and provide an example?